src/radical/pilot/agent/executing/popen.py

```diff
@@ -474,5 +559,236 @@ def _check_running(self):
         return action
 
 
+    # --------------------------------------------------------------------------
+    def _get_check(self, event):
+
+        return '\\\n        || (echo "%s failed"; false) || exit 1\n' % event
+
+
+    # --------------------------------------------------------------------------
+    #
+    # launcher
+    #
+    def _get_prof(self, event, tid, msg=''):
+
+        return '$RP_PROF %s\n' % event
+
+
+    # --------------------------------------------------------------------------
+    #
+    # launcher
+    #
+    def _get_launch_env(self, task, launcher):
+
+        ret  = ''
+        cmds = launcher.get_launcher_env()
+        for cmd in cmds:
+            ret += '%s %s' % (cmd, self._get_check('launcher env'))
+        return ret
+
+
+    # --------------------------------------------------------------------------
+    #
+    def _get_pre_launch(self, task, launcher):
+
+        ret  = ''
+        cmds = task['description']['pre_launch']
+        for cmd in cmds:
+            ret += '%s %s' % (cmd, self._get_check('pre_launch failed'))
+
+        return ret
+
+
+    # --------------------------------------------------------------------------
+    #
+    def _get_launch_cmds(self, task, launcher, exec_path):
+
+
+        ret  = ''
+        cmds = ru.as_list(launcher.get_launch_cmds(task, exec_path))
+        for cmd in cmds:
```

Comment:
all launch methods return a string (one command), should we have just `cmd = launcher.get_launch_cmd(task)`? (and not to have `exec_path` as an attribute there, since we build a command without that argument)

Reply:
I wanted to give the LMs the chance to add commands, like set an env, lookup some endpoint, or whatever.

I don't understand the comment about `exec_path` - it is passed to `launcher.get_launch_cmds(task, exec_path)` as you say, but it is actually used there, and the LM cannot infer the location of the exec script otherwise.  What am I missing?

Reply:
Understand about having multiple commands, but then there is some inconsistency in what you expect as output and what it gives -> currently these methods give `str` as output, but `list` is expected.

With `exec_path` - it is just added at the end of the produced string (not used to make any decisions inside the method and is not modified either), and my thought was (considering method provides a single command) that method generates a command, which is set within the script and represented as a cmd-line while adding a corresponding executable path -> request for a command (`tail -f`) which is used to have a [complete] cmd-line (`cat <file_name>`). Sorry, could sound over-complicated, but also could be skipped if we talk about list of commands (cmd-lines)